File: app/src/main/java/com/situ2001/cliptokindle/util/PageGenerator.java
package com.situ2001.cliptokindle.util;

import androidx.annotation.NonNull;

import com.situ2001.cliptokindle.bean.DisplayableList;

/**
 * A simple tool for generating static web page
 */
public class PageGenerator {
    private static PageGenerator pageGenerator;

    private StringBuilder msg;
    private final DisplayableList displayableList;

    public static PageGenerator getPageGenerator() {
        return pageGenerator;
    }

    private PageGenerator(DisplayableList displayableList) {
        this.displayableList = displayableList;
    }

    /**
     * build a PageGenerator
     * @param list An DisplayList that will be used by PageGenerator
     */
    public synchronized static void build(DisplayableList list) {
        if (pageGenerator == null) {
            pageGenerator = new PageGenerator(list);
        }
    }

    @NonNull
    public String generate() {
        msg = new StringBuilder();

        begin();
        displayableList.forEach(text -> msg.append(text.toHtml()));
        end();

        return msg.toString();
    }

    private void begin() {
        msg.append("<html><body><h1>Clip to Kindle</h1>\n");
        msg.append("<a href=\"/\"><h2>Reload this page</h2></a>\n<br>\n");
        msg.append("<h2>Links</h2>\n");
    }

    private void end() {
        msg.append("</body></html>\n");
    }
}

File: app/src/main/java/com/situ2001/cliptokindle/util/PageGenerator.kt
package com.situ2001.cliptokindle.util

import com.situ2001.cliptokindle.text.Text
import com.situ2001.cliptokindle.text.TextSet

class PageGenerator private constructor(private val textSet: TextSet) {
    fun generate(): String {
        val msg = StringBuilder()
        begin(msg)
        textSet.forEach { msg.append(it.toHtml()) }
        end(msg)
        return msg.toString()
    }

    private fun begin(msg: StringBuilder) {
        msg.append("<html><body><h1>Clip to Kindle</h1>\n")
        msg.append("<a href=\"/\"><h2>Reload this page</h2></a>\n<br>\n")
        msg.append("<h2>Links</h2>\n")
    }

    private fun end(msg: StringBuilder) {
        msg.append("</body></html>\n")
    }

    companion object {
        private var pageGenerator: PageGenerator? = null
        fun getPageGenerator(): PageGenerator? {
            return pageGenerator
        }

        @Synchronized
        fun build(textSet: TextSet) {
            if (pageGenerator == null) {
                pageGenerator = PageGenerator(textSet)
            }
        }
    }
}